src/app/CommandSender.cpp

```diff
@@ -540,13 +540,27 @@ CHIP_ERROR CommandSender::FinishCommand(FinishCommandParameters & aFinishCommand
 CHIP_ERROR CommandSender::AddRequestData(const CommandPathParams & aCommandPath, const DataModel::EncodableToTLV & aEncodable,
                                          AddRequestDataParameters & aAddRequestDataParams)
 {
+    ReturnErrorOnFailure(AllocateBuffer());
+
+    RollbackData rollbackData;
+    rollbackData.Checkpoint(*this);
     PrepareCommandParameters prepareCommandParams(aAddRequestDataParams);
-    ReturnErrorOnFailure(PrepareCommand(aCommandPath, prepareCommandParams));
-    TLV::TLVWriter * writer = GetCommandDataIBTLVWriter();
-    VerifyOrReturnError(writer != nullptr, CHIP_ERROR_INCORRECT_STATE);
-    ReturnErrorOnFailure(aEncodable.EncodeTo(*writer, TLV::ContextTag(CommandDataIB::Tag::kFields)));
-    FinishCommandParameters finishCommandParams(aAddRequestDataParams);
-    return FinishCommand(finishCommandParams);
+    TLV::TLVWriter * writer = nullptr;
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    SuccessOrExit(err = PrepareCommand(aCommandPath, prepareCommandParams));
+    writer = GetCommandDataIBTLVWriter();
+    VerifyOrExit(writer != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    SuccessOrExit(err = aEncodable.EncodeTo(*writer, TLV::ContextTag(CommandDataIB::Tag::kFields)));
+    {
+        FinishCommandParameters finishCommandParams(aAddRequestDataParams);
+        SuccessOrExit(err = FinishCommand(finishCommandParams));
+    }
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        LogErrorOnFailure(rollbackData.Rollback(*this));
```

Comment:
Can we make ~RollbackData roll back automatically unless it's been committed?  And then we don't need the `exit` bits here....

Reply:
I like it. Thanks! Updated let me know if this is more to your liking